allow agents to always reopen tickets, regardless of group.follow_up_possible

martinvonwittich opened the pull request:
As discussed in the community: https://community.zammad.org/t/allow-only-agent-to-re-open-the-ticket-follow-up-possible/1906

This commit changes the behavior of the option group.follow_up_possible
(admin -> manage -> groups -> <group> -> follow up possible). Before this
commit, setting this option to "do not reopen ticket but create a new ticket"
meant that _both_ agents and customers weren't allowed to reopen closed
tickets. This is problematic because it means that an agent who accidentally
closed a ticket has no way to undo this, which I believe wasn't the intention
of this option.

This commit changes the behavior so that agents are now exempt from this
limitation and can close and reopen tickets at whim.

As I said in the community thread, I see no reason to add a separate option for this new behavior, because I believe this is how it's supposed to work; preventing agents from reopening tickets just doesn't seem to be very useful to me. If this is not acceptable, I could of course still change that :)

martinvonwittich committed:
allow agents to always reopen tickets, regardless of group.follow_up_possible

This commit changes the behavior of the option group.follow_up_possible
(admin -> manage -> groups -> <group> -> follow up possible). Before this
commit, setting this option to "do not reopen ticket but create a new ticket"
meant that _both_ agents and customers weren't allowed to reopen closed
tickets. This is problematic because it means that an agent who accidentally
closed a ticket has no way to undo this, which I believe wasn't the intention
of this option.

This commit changes the behavior so that agents are now exempt from this
limitation and can close and reopen tickets at whim.

MrGeneration commented:
Hey Billy,

could you check this when you find some free time please? :)

Thanks!

thorsteneckel commented:
Hi @martinvonwittich - thanks a ton! Can someone (@MrGeneration / @martinvonwittich ) please create an issue for this? We need to have a reference for the CHANGELOG / Release news notes.

Other than that we need to cover the changes with tests. There is already [a request spec](https://github.com/zammad/zammad/blob/develop/spec/requests/ticket_spec.rb) for the controller but I haven't found any example for checking the follow up functionality. However, the spec is in a pretty bad shape so please take inspiration for creating a test from a proper spec like e.g. [this one](https://github.com/zammad/zammad/blob/develop/spec/requests/time_accounting_spec.rb). Let me know if you have any questions.

martinvonwittich commented:
Done: #2534

martinvonwittich committed:
allow agents to always reopen tickets: fixed spec (refs: #2534)

- Removed the old test and replaced it with one that's hopefully more concise
  and readable.
- Added tests for the follow_up_possible = yes behavior.
- Used a helper function to ensure that no code is duplicated.

martinvonwittich commented:
> Other than that we need to cover the changes with tests. There is already a request spec for the controller but I haven't found any example for checking the follow up functionality.

Oh, there actually was a test for this, and my change broke it :D

```
  1) Ticket request handling does ticket with follow up possible set to new_ticket (06.01)
     Failure/Error: expect(response).to have_http_status(422)
       expected the response to have status code 422 but it was 200
     # ./spec/requests/ticket_spec.rb:1912:in `block (3 levels) in <top (required)>'
```

Unfortunately it had a lot of duplicated code and it also didn't actually test the `follow_up_possible = yes` behavior, so I decided to remove it completely and wrote a new one. It took me some hours to figure out a clean way to prevent code duplication; in the end I decided on a helper function so I don't have to repeat all those `expect` lines for each test. I hope that is acceptable in specs...?

I moved the new test into its own `describe '/api/v1/tickets'` section; I assume that the giant `describe 'request handling'` is one of the things that's bad style in the spec file?

Here's the new test: https://github.com/zammad/zammad/pull/2531/commits/4018d187f76edf568dea610f1598e319ea860b7a

Output:

```
zammad@martin ~ (agents-can-always-reopen-tickets) % bundle exec bin/rspec spec/requests/ticket_spec.rb -e '/api/v1/tickets'
Running via Spring preloader in process 9254
Run options: include {:full_description=>/\/api\/v1\/tickets/}

Ticket
  /api/v1/tickets
    follow_up_possible
      when follow_up_possible = yes
        allows admins to reopen tickets
        allows agents to reopen tickets
        allows customers to reopen tickets
      when follow_up_possible = new_ticket
        allows admins to reopen tickets
        allows agents to reopen tickets
        does not allow customers to reopen tickets

Finished in 4.13 seconds (files took 1.05 seconds to load)
6 examples, 0 failures
```

thorsteneckel commented:
Beautiful ❤️ the helper method is a step in the right direction we can easily migrate it over to a [RSpec shared example](https://relishapp.com/rspec/rspec-core/docs/example-groups/shared-examples) like we use it in [other specs](https://github.com/zammad/zammad/blob/57b17f7b81676eb32185905edd918dfd6723fdb8/spec/models/object_manager/attribute/validation/future_past_spec.rb#L19-L41).

I'll talk to the team and let you know.

martinvonwittich commented:
Dammit, tried to rename the branch in my fork from `agents-can-always-reopen-tickets` to `2534-agents-can-always-reopen-tickets` now that I have a proper bug number, and that closed the merge request :/

thorsteneckel commented:
Hi @martinvonwittich - thanks again for creating a PR for this. Unfortunately we had to work on the spec which leads in our current workflow (GitHub PR -> review/CI on internal GitLab -> GitHub) to overwriting you as an author of the commit :/ I hope that's not too demotivating. We're still searching for a way to keep the original author of the code...
